core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java

```diff
@@ -258,21 +261,28 @@ public T responseCauseSanitizer(Function<? super Throwable, ?> responseCauseSani
     }
 
     /**
-     * Sets the rate at which to sample requests to log. Any number between {@code 0.0} and {@code 1.0} will
-     * cause a random sample of the requests to be logged. The random sampling is appropriate for low-traffic
-     * (ex servers that each receive &lt;100K requests). If unset, all requests will be logged.
+     * Sets the {@link Sampler} that determines which request needs logging.
      */
-    public T samplingRate(float samplingRate) {
-        checkArgument(0.0 <= samplingRate && samplingRate <= 1.0, "samplingRate must be between 0.0 and 1.0");
-        this.samplingRate = samplingRate;
+    public T sampler(Sampler<? super RequestContext> sampler) {
```

Comment:
I guess a user cannot specify `Sampler<ClientRequestContext` or `Sampler<ServiceRequestContext>` because they are not matched by `Sampler<? super RequestContext>`. How about pushing this method down to `LoggingClientBuilder` and `LoggingServiceBuilder` so that there are two different methods:

- `LoggingClientBuilder.sampler(Sampler<? super ClientRequestContext>)`
- `LoggingServiceBuilder.sampler(Sampler<? super ServiceRequestContext>)`

You'll push down the `sampler` field and `sampler()` getter method as well.

Reply:
Thank you for review. I was mistaken that I was using 'Sampler<? extends RequestContext>'.
Fixed in 8994cfb